runtime-jvm/main/src/main/scala/util/Critbyte.scala

```diff
@@ -138,6 +148,55 @@ object Critbyte {
 
     lazy val prefix = smallestKey.take(critbyte)
 
+    def unionWith(b: Critbyte[A])(f: (A, A) => A) = b match {
+      case l@Leaf(_) => l.unionWith(this)(f)
+      case Branch(cb, sk, r, ch) =>
+        def worstCase =
+          runt.unionWith(r)(f).unionWith(children.foldLeft(b) { (acc, c) =>
+            acc.unionWith(c)(f)
+          })(f)
+        try {
+          val sdi = sk smallestDifferingIndex smallestKey
+          if (sdi < critbyte && sdi < cb)
+            // The union has a new, shorter prefix
+            Branch(sdi, smallestKey min sk, emptyLeaf,
+                   emptyChildArray[A].updated(unsigned(smallestKey(sdi)), this)
+                                     .updated(unsigned(sk(sdi)), b))
+          else if (critbyte < sdi && sdi <= cb || critbyte == sdi && sdi < cb)
+            // The whole tree `b` belongs under one of this branch's children
+            copy(smallestKey = smallestKey min sk,
+                 children =
+                   children.updated(critbyte,
+                                    children(critbyte).unionWith(b)(f)))
+          else if (cb < sdi && sdi <= critbyte || cb == sdi && sdi < critbyte)
+            // This whole branch belongs under one of the children of `b`
+            copy(smallestKey = smallestKey min sk,
+                 children =
+                   children.updated(cb,
+                                    children(cb).unionWith(this)(f)))
+          else if (sdi >= critbyte && critbyte == cb) {
+            // The two trees have the same prefix
+            val newChildren = emptyChildArray[A].clone
+            0 until children.size foreach { i =>
+              newChildren(i) = children(i).unionWith(ch(i))(f)
+            }
+            val (newRunt, otherRunt) =
+              if ((sk min smallestKey) == sk) (r, runt) else (runt, r)
+            val newSmallest = sk min smallestKey
+            Branch(critbyte,
+                   newSmallest,
+                   newRunt,
+                   newChildren).unionWith(otherRunt)(f)
+          }
+          else
+            // The sdi of smallest keys is after both critbytes, telling us nothing.
+            worstCase
+        } catch { case Bytes.Seq.NotFound() =>
+          // The smallest key of both trees is the same
+          worstCase
```

Comment:
We can probably expand `worstCase` to compute the prefixes and then descend into one of the other cases if there's a common prefix.